libparsec/crates/client/src/certif/shamir_recovery_setup.rs

```diff
@@ -0,0 +1,448 @@
+    }
+    // 3. Check recipients status
+
+    let mut participants_id: HashSet<_> = share_recipients.keys().collect();
+    participants_id.insert(&author_id);
+    // TODO implement a get_users method instead of list_user (let the DB do its job)
+    let participants = certificate_ops.list_users(true, None, None).await?;
+    let participants = participants
+        .iter()
+        .filter(|info| participants_id.contains(&info.id));
+
+    // no participant is missing
+    for &&id in &participants_id {
+        if !participants.clone().any(|info| info.id == id) {
+            return Err(CertifShamirError::MissingUser(id));
+        }
+    }
+
+    // no participant is revoked
+    for info in participants {
+        if info.revoked_on.is_some() {
+            return Err(CertifShamirError::UserRevoked(info.id));
+        }
+    }
+    // 4. Generate certificates
+
+    let brief: Bytes = ShamirRecoveryBriefCertificate {
+        author: author_device,
+        timestamp,
+        user_id: author_id,
+        threshold: NonZeroU64::new(threshold.into()).expect("shamir threshold is zero"), // checked during the first step
+        per_recipient_shares: share_recipients
+            .iter()
+            .map(|(&k, &v)| (k, NonZeroU64::new(v.into()).expect("Share count is zero")))
+            .collect(),
+    }
+    .dump_and_sign(&certificate_ops.device.signing_key)
+    .into();
+
+    let recovery_device = LocalDevice::generate_new_device(
+        certificate_ops.device.organization_addr.clone(),
+        certificate_ops.device.initial_profile,
+        certificate_ops.device.human_handle.clone(),
+        DeviceLabel::try_from(
+            format!(
+                "recovery-{timestamp}-{}",
+                u32::from_be_bytes(brief[..4].try_into().expect("u32 from 4 u8 should be ok"))
+            )
+            .as_str(),
+        )
+        .expect("Invalid device label"),
+        Some(certificate_ops.device.user_id),
+        None,
+        None,
+        None,
+    );
+
+    let data_key = SecretKey::generate();
+    let reveal_token = InvitationToken::default();
+
+    let ciphered_data = data_key.encrypt(&recovery_device.dump()).into();
+
+    let shark_shares = shamir_make_shares(
+        threshold,
+        &ShamirRecoverySecret {
+            data_key,
+            reveal_token,
+        }
+        .dump()?,
+        total_share_count.into(),
+    );
+    let mut idx = 0_usize;
+    let mut shares = Vec::new();
+    for (&share_recipient_id, &share_count) in share_recipients {
+        let pub_key = &certificate_ops
+            .store
+            .for_read(|store| store.get_user_certificate(UpTo::Current, share_recipient_id))
+            .await??
+            .public_key;
+        let ciphered_share = ShamirRecoveryShareData {
+            weighted_share: shark_shares[idx..idx + share_count as usize].to_vec(),
+        }
+        .dump_sign_and_encrypt_for(&certificate_ops.device.signing_key, pub_key);
```

Comment:
Signing is not required here since the certificate is already signed, but maybe it's more convenient to keep it?